# Анализ базы данных средствами SQL
## Задача:
Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных. В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

## Описание датасетов

### Таблица books

Содержит данные о книгах:

- book_id — идентификатор книги;
- author_id — идентификатор автора;
- title — название книги;
- num_pages — количество страниц;
- publication_date — дата публикации книги;
- publisher_id — идентификатор издателя.

### Таблица authors

Содержит данные об авторах:

- author_id — идентификатор автора;
- author — имя автора.

### Таблица publishers

Содержит данные об издательствах:

- publisher_id — идентификатор издательства;
- publisher — название издательства;

### Таблица ratings

Содержит данные о пользовательских оценках книг:

- rating_id — идентификатор оценки;
- book_id — идентификатор книги;
- username — имя пользователя, оставившего оценку;
- rating — оценка книги.

### Таблица reviews

Содержит данные о пользовательских обзорах на книги:

- review_id — идентификатор обзора;
- book_id — идентификатор книги;
- username — имя пользователя, написавшего обзор;
- text — текст обзора.



## Исследование таблиц

In [1]:
import pandas as pd
from sqlalchemy import create_engine

db_config = {'user': 'praktikum_student', # имя пользователя
            'pwd': 'Sdf4$2;d-d30pp', # пароль
            'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
            'port': 6432, # порт подключения
            'db': 'data-analyst-final-project-db'} # название базы данных

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                db_config['pwd'],
                                                db_config['host'],
                                                db_config['port'],
                                                db_config['db'])

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

### Таблица books

In [2]:
query = '''
            SELECT * 
            FROM books
        '''
books = pd.io.sql.read_sql(query, con = engine)
books.head()

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [3]:
books.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


### Таблица authors

In [4]:
query = '''
            SELECT * 
            FROM authors
        '''
authors = pd.io.sql.read_sql(query, con = engine)
authors.head()

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [5]:
authors.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


### Таблица publishers

In [6]:
query = '''
            SELECT * 
            FROM publishers
        '''
publishers = pd.io.sql.read_sql(query, con = engine)
publishers.head()

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [7]:
publishers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


### Таблица ratings

In [8]:
query = '''
            SELECT * 
            FROM ratings
        '''
ratings = pd.io.sql.read_sql(query, con = engine)
ratings.head()

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [9]:
ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


### Таблица reviews

In [10]:
query = '''
            SELECT * 
            FROM reviews
        '''
reviews = pd.io.sql.read_sql(query, con = engine)
reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


## Ответы на вопросы

### Сколько книг вышло после 1 января 2000 года?

In [11]:
query = '''
            SELECT COUNT(book_id) count_of_books
            FROM books 
            WHERE publication_date > '01-01-2000'
        '''
pd.io.sql.read_sql(query, con = engine)

count_of_books
0             819

После 1 января 2000 года было выпущено 819 книг.

### Для каждой книги посчитайте количество обзоров и среднюю оценку


Для каждой книги посчитаем количество обзоров и среднюю оценку. Затем отсортируем полученные результаты по количеству обзоров и по средней оценке соответственно:

In [12]:
query = '''
SELECT
    books.title,
    COUNT (DISTINCT review_id) AS amount_review,
    AVG(rating) AS average_rating
FROM books
LEFT JOIN reviews ON reviews.book_id = books.book_id
LEFT JOIN ratings ON ratings.book_id = books.book_id
GROUP BY books.title
ORDER BY COUNT(reviews.review_id) DESC
'''

pd.io.sql.read_sql(query, con = engine)

title  amount_review  \
0                              Twilight (Twilight  #1)              7   
1                  The Hobbit  or There and Back Again              6   
2                               The Catcher in the Rye              6   
3    Harry Potter and the Prisoner of Azkaban (Harr...              6   
4    Harry Potter and the Chamber of Secrets (Harry...              6   
..                                                 ...            ...   
994  Disney's Beauty and the Beast (A Little Golden...              0   
995                          Essential Tales and Poems              0   
996                               Leonardo's Notebooks              0   
997                            The Natural Way to Draw              0   
998   The Cat in the Hat and Other Dr. Seuss Favorites              0   

     average_rating  
0          3.662500  
1          4.125000  
2          3.825581  
3          4.414634  
4          4.287500  
..              ...  
994        4.000000  
995        4.000000  
996        4.000000  
997        3.000000  
998        5.000000  

[999 rows x 3 columns]

In [13]:
query = '''
SELECT
    books.title,
    COUNT (DISTINCT review_id) AS amount_review,
    AVG(rating) AS average_rating
FROM books
LEFT JOIN reviews ON reviews.book_id = books.book_id
LEFT JOIN ratings ON ratings.book_id = books.book_id
GROUP BY books.title
ORDER BY average_rating DESC
'''

pd.io.sql.read_sql(query, con = engine)

title  amount_review  \
0                 Pop Goes the Weasel (Alex Cross  #5)              2   
1                                          Angels Fall              2   
2                 Piercing the Darkness (Darkness  #2)              2   
3     The Cat in the Hat and Other Dr. Seuss Favorites              0   
4                             Neil Gaiman's Neverwhere              2   
..                                                 ...            ...   
994  The World Is Flat: A Brief History of the Twen...              3   
995                                              Junky              2   
996                                      Drowning Ruth              3   
997                  His Excellency: George Washington              2   
998                               Harvesting the Heart              2   

     average_rating  
0              5.00  
1              5.00  
2              5.00  
3              5.00  
4              5.00  
..              ...  
994            2.25  
995            2.00  
996            2.00  
997            2.00  
998            1.50  

[999 rows x 3 columns]

Строк стало меньше, похоже, что в названиях книг закрался дубликат.

In [14]:
query = '''
SELECT
    title,
    COUNT(title)
FROM books
GROUP BY title
HAVING COUNT(title) > 1
'''

pd.io.sql.read_sql(query, con = engine)

title  count
0  Memoirs of a Geisha      2

Дубликат обнаружен.

Мы посчитали количество обзоров и среднюю оценку для каждой книги. Видим, что наибольшее количество обзоров у книги Twilight (Twilight #1) - 7. Наивысшая средняя оценка - 5, а самая низкая - 1.50.

### Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры

In [15]:
query = '''
            SELECT b.publisher_id, p.publisher, COUNT(*) count_books
            FROM books b
            LEFT JOIN publishers p ON b.publisher_id = p.publisher_id
            WHERE num_pages > 50
            GROUP BY b.publisher_id, p.publisher
            ORDER BY COUNT(*) DESC
            LIMIT 1
        '''
pd.io.sql.read_sql(query, con = engine)

publisher_id      publisher  count_books
0           212  Penguin Books           42

Издательство "Penguin Books" выпустило 42 книги толще 50 страниц. Это около 4% от всех книг в данных.


### Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [16]:
query = '''
            SELECT a.author, AVG(ratings)
            FROM (
            SELECT book_id, AVG(rating) ratings
            FROM ratings
            GROUP BY book_id
            HAVING COUNT(*) > 50
            ) rat
            JOIN (SELECT book_id, author_id FROM books) b ON rat.book_id=b.book_id
            JOIN (SELECT author_id, author FROM authors) a ON b.author_id=a.author_id
            GROUP BY a.author
            ORDER BY avg(ratings) DESC

            
        '''
pd.io.sql.read_sql(query, con = engine)

author       avg
0                          J.K. Rowling/Mary GrandPré  4.283844
1                   Markus Zusak/Cao Xuân Việt Khương  4.264151
2                                      J.R.R. Tolkien  4.258446
3                                   Louisa May Alcott  4.192308
4                                        Rick Riordan  4.080645
5                                     William Golding  3.901408
6                                       J.D. Salinger  3.825581
7            Paulo Coelho/Alan R. Clarke/Özdemir İnce  3.789474
8   William Shakespeare/Paul Werstine/Barbara A. M...  3.787879
9                                           Dan Brown  3.754540
10                                         Lois Lowry  3.750000
11            George Orwell/Boris Grabnar/Peter Škerl  3.729730
12                                    Stephenie Meyer  3.662500
13                                     John Steinbeck  3.622951

Автор с самой высокой оценкой - Джоан Роулинг, автор книг о Гарри Поттере. Тут она представлена в соавторстве с Мэри Гранпрэ(иллюстратором). Что не удивительно, потому что книги о Гарри Поттере одни из самых популярных в мире среди всех возрастных категорий.

### Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [17]:
query = '''
            SELECT AVG(count_reviews)
            FROM (
            SELECT username, COUNT(*) count_reviews
            FROM reviews
            WHERE username IN
            (
            SELECT username--, COUNT(*)
            FROM ratings
            GROUP BY username
            HAVING COUNT(*) > 50
            )
            GROUP BY username) rev
        

        '''
pd.io.sql.read_sql(query, con = engine)

avg
0  24.333333

Среднее количество обзоров от пользователей, которые поставили более 50 оценок - 24.

## Общий итог:

Мы проанализировали базу данных в соответствии с установленным ходом работы. Импортировали библиотеки, подключились к базе данных и провели первичное ознакомление с данными. Последовательно решили все необходимые задачи в рамках цели работы, сопровождая их соответствующими выводами.

В базе данных содержатся данные о 1000 книгах, об 636 авторах, об 340 издательствах, о 6456 пользовательских оценках книг и о 2793 пользовательских обзорах.

В итоге по данным, выяснили:
- После 1 января 2000 года было выпущено 819 книг.
- Наибольшее количество обзоров у книги Twilight (Twilight #1) - 7. 
- Наивысшая средняя оценка - 5, а самая низкая - 1.50.
- Издательство "Penguin Books" выпустило 42 книги толще 50 страниц. Это около 4% от всех книг в данных.
- Автор с самой высокой оценкой - Джоан Роулинг, автор книг о Гарри Поттере.
- Среднее количество обзоров от пользователей, которые поставили более 50 оценок - 24.